arch/arm/core/aarch32/cortex_m/fault_s.S

```diff
@@ -76,6 +76,53 @@ SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_debug_monitor)
 #endif /* CONFIG_ARMV6_M_ARMV8_M_BASELINE */
 SECTION_SUBSEC_FUNC(TEXT,__fault,z_arm_exc_spurious)
 
+#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
```

Comment:
I would be really fond of a single code flow, with in-lining the instructions needed for CONFIG_EXTRA_EXCEPTION_INFO

What you gain is
- you do not need to have #else code, i.e. !CONFIG_EXTRA_EXCEPTION_INFO, so we make it clear that the option only adds extra code.
- you do not need to recover the MSP in the Baseline code path

Something along the following:

```
    mrs  r0, MSP
    mrs  r1, PSP
    push {r0, lr}
#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
    mov  r3, r11
    mov  r2, r10
    push {r2, r3}
    mov  r3, r9
    mov  r2, r8
    push {r2, r3}
    push {r4-r7}
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
    push {r4-r11}
#endif
    mrs  r3, MSP /* pointer to the dumped callee-saved registers in main stack */
#endif /* CONFIG_EXTRA_EXCEPTION_INFO */
    mov r2, lr /* EXC_RETURN */
    bl z_arm_fault
#if defined(CONFIG_EXTRA_EXCEPTION_INFO)
    /* Restore dumped called saved registers */
#if defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
    pop { r4-r7 }
    pop { r0-r3 }
    mov r11, r3
    mov r10, r2
    mov r9, r1
    mov r8, r0
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
    pop {r4-r11}
#endif
#endif /* CONFIG_EXTRA_EXCEPTION_INFO */
```

Reply:
Actually, @chrisc11 I am wondering whether we need to "restore" the callee-saved registers, here. Are they not preserved by the ARM procedure calling standard conventions? Could we just pop them, to restore the stack pointer?